README.md

```diff
@@ -1,26 +1,230 @@
-# reagents — Composable lock-free data and synchronization structures
+[API reference](https://ocaml-multicore.github.io/reagents/doc/)
 
-reagents library is in experimental stage. It is distributed under the 
-ISC license.
+# Reagents — Composable lock-free data and synchronization structures
 
-Homepage: https://github.com/ocaml-multicore/reagents
+Reagents are an experimental library for writing multicore programs. Reagents
+provide an expressive framework for composable multithreading. They support both
+fine- and coarse-grained multithreading and incorporate mechanisms for efficient
+retrying.
 
-## Installation
+## Contents
 
-The library is not in opam at the moment. But do feel invited to hack! 
+- [Motivation](#motivation)
+- [Limitations](#limitiations)
+- [Getting Reagents](#getting-reagents)
+- [Key Concepts](#key-concepts)
+- [Development](#development)
+- [License](#license)
+- [Reading](#reading)
+
+## Motivation
+
+Reagents strive to be a comprehensive framework for all things concurrent and
+parallel. In particular, they have the following advantages over traditional
+approaches:
+
+- _Composability_. Operations are trivial to compose with the set of provided
+  composition operators. An item can be moved from one lock-free stack to
+  another in a single atomic lock-free step. Further, a release of one lock can
+  be melded with acquisition of another one. This has far-reaching consequences.
+  For example, a simple way to implement an LRU cache is using a queue and a
+  map, but traditional implementations of thread-safe queue and map do not help
+  as both have to be updated at once. Reagents make that trivial (in contrast
+  with implementing thread-safe LRU cache from scratch) and let us maintain
+  useful properties (in contrast with surrounding both updates with a lock).
+
+- _Expressiveness_. Reagents provide building blocks for various multithreading
+  patterns: communication via sharing memory and message passing, active and
+  passive invocation of operations, conjuction (pair) and disjunction (choice)
+  of operations.
+
+- _Fine-grained multithreading_. Low-level synchronisation primitives tend to
+  perform and scale better than high-level ones. Reagents use fine-grained
+  multithreading internally and expose them for expert users.
+
+- _Efficient retrying_. Reagents parametrise over scheduler to suspend and
+  resume fibers as the conditions for their progress are meet or not. This makes
+  common anti-patterns such as busy-waiting easy to avoid.
+
+## Limitiations
+
+Reagents are weaker than transactional memory. A reagent must be decomposable
+into a list of compare-and-swap operations. This eliminates the need for global
+serializability of transactions.
```

Comment:
Hmm... I'm not sure what would be the best way to phrase this, but I feel this might give wrong impressions.  I mean, transactional memory can be disjoint access parallel, require linearizability (rather than serializability), or be e.g. based on MVCC (like in Clojure) and offer various levels of isolation.

Reply:
I see, how about something vaguer, e.g. "This eliminates the need for any extra accounting for performed k-cases"?

Note to self: add "on distinct locations" after "compare and swap ops".

Reply:
I spent some time reading last night about related terminology (it seems I need to refresh my memory on these every now and then) and I think k-CAS itself could be described as "strictly serializable" which means that it is both linearizable and serializable, which is about as strong as it gets.  And Reagents inherit this property for the composable accesses.  But I digress.

The way Turon seems to express it is that the idea with STM is that one just wraps an imperative algorithm inside an atomic block with all the shared memory accesses inside of it.

With STM all shared memory accesses (reads and writes) can be further composed and performed atomically.  The advantage of this is that it makes programming easy.  The disadvantage is that it is often possible to do better.

With Reagents the idea is then to perform only a portion of the atomic accesses in a composable manner.  Accesses can be performed before those composable accesses and there is also support to perform actions after such composable accesses are committed.  This puts the onus on the programmer to ensure composability still holds.

Reply:
Hmm... So, the "limitation" or, perhaps disadvantage, of Reagents is that the API makes it easier to express incorrect algorithms.  Perhaps something like this:

Use of fine-grained multithreading features inside reagents is not automatically thread safe and requires careful analysis to ensure correctness.

But, yeah, feel free to choose how to express this.  At any rate, I think Reagents sans the unsafe accesses are at least as strong as transactional memory.